src/lib/merkle_ledger/database.ml

```diff
@@ -700,6 +700,58 @@ module Make (Inputs : Inputs_intf) :
     List.map2_exn dependency_dirs dependency_hashes ~f:(fun dir hash ->
         Direction.map dir ~left:(`Left hash) ~right:(`Right hash) )
 
+  let merkle_path_batch mdb locations =
+    let locations =
+      List.map locations ~f:(fun location ->
+          if Location.is_account location then
+            Location.Hash (Location.to_path_exn location)
+          else (
+            assert (Location.is_hash location) ;
+            location ) )
+    in
+    let rev_locations, rev_directions, rev_lengths =
+      let rec loop locations loc_acc dir_acc length_acc =
+        match (locations, length_acc) with
+        | [], _ :: length_acc ->
+            (loc_acc, dir_acc, length_acc)
+        | k :: locations, length :: length_acc ->
+            if Location.height ~ledger_depth:mdb.depth k >= mdb.depth then
+              loop locations loc_acc dir_acc (0 :: length :: length_acc)
+            else
+              let sibling = Location.sibling k in
+              let sibling_dir =
+                Location.last_direction (Location.to_path_exn k)
+              in
+              loop
+                (Location.parent k :: locations)
+                (sibling :: loc_acc) (sibling_dir :: dir_acc)
+                ((length + 1) :: length_acc)
+        | _ ->
+            assert false
+      in
+      loop locations [] [] [ 0 ]
+    in
+    let rev_hashes = get_hash_batch mdb rev_locations in
+    let rec loop directions hashes lengths acc =
+      match (directions, hashes, lengths, acc) with
+      | [], [], [], _ (* actually [] *) :: acc_tl ->
+          acc_tl
+      | _, _, 0 :: lengths, _ ->
+          loop directions hashes lengths ([] :: acc)
+      | ( direction :: directions
+        , hash :: hashes
+        , length :: lengths
+        , acc_hd :: acc_tl ) ->
+          let dir =
+            Direction.map direction ~left:(`Left hash) ~right:(`Right hash)
```

Comment:
Shouldn't these be flipped? It appears from the function above that we are returning the directions we traverse, and then returning the locations to the sibling hashes. So I would think we want to flip the directions here in order to build the proper merkle path.

Reply:
Merkle path convention is that in each element there is a direction to the node itself (from its parent) and hash of sibling.

Because `Location.last_direction (Location.to_path_exn k)`, `Direction.map direction ~left:(Left hash) ~right:(Right hash)` returns direction to the node itself with hash of a sibling, following the convention.